gsy_myco_sdk/matching_algorithms/preferred_partners_algorithm.py

```diff
@@ -179,3 +125,99 @@ def _get_required_energy_and_rate_from_order(
+        already selected order energy in order to be able to reuse the same order in future
+        matches.
+        """
+        if bid["id"] not in available_order_energy:
+            available_order_energy[bid["id"]] = bid["energy"]
+        if offer["id"] not in available_order_energy:
+            available_order_energy[offer["id"]] = offer["energy"]
+        available_offer_energy = available_order_energy[offer["id"]]
+        available_bid_energy = available_order_energy[bid["id"]]
+
+        for index, bid_requirement in enumerate(bid.get("requirements") or []):
+            bid_required_energy, bid_required_clearing_rate = (
+                cls._get_required_energy_and_rate_from_order(bid, bid_requirement))
+            preferred_offers = []
+            for partner in bid_requirement.get("trading_partners") or []:
+                preferred_offers.extend(offers_mapping.get(partner) or [])
+
+            for offer_requirement in offer.get("requirements") or [{}]:
+                if not cls._can_order_be_matched(
+                        bid, offer, bid_requirement, offer_requirement):
+                    continue
+
+                offer_required_energy, offer_required_clearing_rate = (
+                    cls._get_required_energy_and_rate_from_order(offer, offer_requirement))
+
+                if (
+                        offer_required_clearing_rate - bid_required_clearing_rate
+                ) > FLOATING_POINT_TOLERANCE:
+                    return None
+
+                selected_energy = min(offer_required_energy, bid_required_energy)
+
+                if (selected_energy <= FLOATING_POINT_TOLERANCE or
+                        available_offer_energy < selected_energy or
+                        available_bid_energy < selected_energy):
+                    return None
+
+                recommendation = BidOfferMatch(
+                    market_id=market_id,
+                    time_slot=time_slot,
+                    bid=bid, offer=offer,
+                    selected_energy=selected_energy,
+                    trade_rate=bid_required_clearing_rate,
+                    matching_requirements={
+                        "bid_requirement": bid_requirement,
+                        "offer_requirement": offer_requirement
+                    })
+                available_order_energy[bid["id"]] -= selected_energy
+                available_order_energy[offer["id"]] -= selected_energy
+                assert all(v >= -FLOATING_POINT_TOLERANCE
+                           for v in available_order_energy.values())
+
+                if bid_requirement.get("energy") is not None:
+                    updated_requirement = deepcopy(bid_requirement)
+                    updated_requirement["energy"] -= selected_energy
+                    assert updated_requirement["energy"] >= -FLOATING_POINT_TOLERANCE
+                    bid["requirements"][index] = updated_requirement
```

Comment:
Can you please explain why you have to do this detour here and not reduce the requirement directly like:
```python
bid["requirements"][index]["energy"] -= selected_energy
```
Thanks!

Reply:
Sure,
```python
recommendation = BidOfferMatch(
                    market_id=market_id,
                    time_slot=time_slot,
                    bid=bid, offer=offer,
                    selected_energy=selected_energy,
                    trade_rate=bid_required_clearing_rate,
                    matching_requirements={
                        "bid_requirement": bid_requirement,
                        "offer_requirement": offer_requirement
                    })
```
A few lines above the one you mentioned, I'm passing the selected bid_requirement to the BidOfferMatch instance (we use this dictionary for validation on the gsy-e side).
So the reason why I'm not mutating the original dict here is to not mess with the one above.
However, I'm starting to feel that I could use `deepcopy` here:
```python
matching_requirements={
                        "bid_requirement": deepcopy(bid_requirement),
                        "offer_requirement": deepcopy(offer_requirement)
                    })
``` 
In this way, I can avoid the detour and therefore simplify the code, let me know if you agree

Reply:
Ahh, yes. This makes sense now.
I agree with your proposal. Only one remark: AFAIR; `copy` is much more lightweight than `deepcopy`, especially in memory usage. And as the `bid_requirement` and `offer_requirement` dicts will probably never be deeply nested, it should work fine here, too, right?

Reply:
@hannesdiedrich the only concern here is the `trading_partners` which is a list and can be mutated, we're not changing anything here but I think we better be safe and keep the deepcopy, do you agree?

Reply:
ok, agreed. Let's resume this discussion if we have a memory leak then.